pkg/querier/ingester_querier.go

```diff
@@ -23,61 +29,98 @@ type IngesterQueryClient interface {
 		if err != nil {
 			return nil, err
 		}
+		return client.(IngesterQueryClient), nil
+	}, replicationSet, f)
+}
 
-		resp, err := f(ctx, client.(IngesterQueryClient))
-		if err != nil {
-			return nil, err
-		}
+func (q *Querier) selectTreeFromIngesters(ctx context.Context, req *querierv1.SelectMergeStacktracesRequest) (*phlaremodel.Tree, error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "SelectTree Ingesters")
+	defer sp.Finish()
+	profileType, err := phlaremodel.ParseProfileTypeSelector(req.ProfileTypeID)
+	if err != nil {
+		return nil, connect.NewError(connect.CodeInvalidArgument, err)
+	}
+	_, err = parser.ParseMetricSelector(req.LabelSelector)
+	if err != nil {
+		return nil, connect.NewError(connect.CodeInvalidArgument, err)
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-		return responseFromIngesters[T]{ingester.Addr, resp}, nil
+	responses, err := forAllIngesters(ctx, q.ingesterQuerier, func(ctx context.Context, ic IngesterQueryClient) (clientpool.BidiClientMergeProfilesStacktraces, error) {
+		return ic.MergeProfilesStacktraces(ctx), nil
 	})
 	if err != nil {
-		return nil, err
+		return nil, connect.NewError(connect.CodeInternal, err)
 	}
-
-	responses := make([]responseFromIngesters[T], 0, len(results))
-	for _, result := range results {
-		responses = append(responses, result.(responseFromIngesters[T]))
+	// send the first initial request to all ingesters.
+	g, gCtx := errgroup.WithContext(ctx)
+	for _, r := range responses {
+		r := r
+		g.Go(util.RecoverPanic(func() error {
+			return r.response.Send(&ingestv1.MergeProfilesStacktracesRequest{
+				Request: &ingestv1.SelectProfilesRequest{
+					LabelSelector: req.LabelSelector,
+					Start:         req.Start,
+					End:           req.End,
+					Type:          profileType,
+				},
+				MaxNodes: req.MaxNodes,
+				// TODO(kolesnikovae): Max stacks.
+			})
+		}))
+	}
+	if err = g.Wait(); err != nil {
+		return nil, connect.NewError(connect.CodeInternal, err)
 	}
 
-	return responses, err
+	// merge all profiles
+	return selectMergeTree(gCtx, responses)
+}
+
+func (q *Querier) selectSeriesFromIngesters(ctx context.Context, req *ingesterv1.MergeProfilesLabelsRequest) ([]ResponseFromReplica[clientpool.BidiClientMergeProfilesLabels], error) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "SelectSeries Ingesters")
+	defer sp.Finish()
+	responses, err := forAllIngesters(ctx, q.ingesterQuerier, func(ctx context.Context, ic IngesterQueryClient) (clientpool.BidiClientMergeProfilesLabels, error) {
+		return ic.MergeProfilesLabels(ctx), nil
+	})
+	if err != nil {
+		return nil, connect.NewError(connect.CodeInternal, err)
+	}
+	// send the first initial request to all ingesters.
+	g, _ := errgroup.WithContext(ctx)
+	for _, r := range responses {
+		r := r
+		g.Go(util.RecoverPanic(func() error {
+			return r.response.Send(req)
+		}))
+	}
+	if err := g.Wait(); err != nil {
+		return nil, connect.NewError(connect.CodeInternal, err)
+	}
+	return responses, nil
```

Comment:
I must be missing something. Here and in some other places we first collect responses from all the clients (ingesters) and only then send the request – can it be done immediately for each client (ingester), i.e in the `forAllIngesters` callback